llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -4189,27 +4184,12 @@ ElementCount LoopVectorizationCostModel::getMaximizedVFForTarget(
         ComputeScalableMaxVF);
     MaxVectorElementCountMaxBW = MinVF(MaxVectorElementCountMaxBW, MaxSafeVF);
 
-    // Collect all viable vectorization factors larger than the default MaxVF
-    // (i.e. MaxVectorElementCount).
-    SmallVector<ElementCount, 8> VFs;
-    for (ElementCount VS = MaxVectorElementCount * 2;
-         ElementCount::isKnownLE(VS, MaxVectorElementCountMaxBW); VS *= 2)
-      VFs.push_back(VS);
-
-    // For each VF calculate its register usage.
-    auto RUs = calculateRegisterUsage(VFs);
-
-    // Select the largest VF which doesn't require more registers than existing
-    // ones.
-    for (int I = RUs.size() - 1; I >= 0; --I) {
-      const auto &MLU = RUs[I].MaxLocalUsers;
-      if (all_of(MLU, [&](decltype(MLU.front()) &LU) {
-            return LU.second <= TTI.getNumberOfRegisters(LU.first);
-          })) {
-        MaxVF = VFs[I];
-        break;
-      }
-    }
+    // Set the max VF to the largest viable vectorization factor less than or
+    // equal to the max vector element count.
+    if (ElementCount::isKnownLE(MaxVectorElementCount,
+                                MaxVectorElementCountMaxBW))
```

Comment:
Is this condition ever false?

Reply:
There is one case where it is. See this thread for context: https://github.com/llvm/llvm-project/pull/132190#discussion_r2014209907

Reply:
That's because of a bug in `LoopVectorizationCostModel::getSmallestAndWidestTypes` where `MinWidth` is initialised with `-1U`, but never set to `MaxWidth` in the `if (ElementTypesInLoop.empty() && !Legal->getReductionVars().empty()) { ... }` block. This results in MaxVectorElementCountMaxBW being 0, since it divides the vector size by a very high number.

If you fix that (best done in a separate PR), you can remove this `if` condition.